src/request-filtering-agent.ts

```diff
@@ -74,112 +77,126 @@ const validateIPAddress = (
         return socket;
-    };
-    return agent;
-}
-
-/**
- * A subclass of http.Agent with request filtering
- */
-export class RequestFilteringHttpAgent extends http.Agent {
-    constructor(options?: http.AgentOptions & RequestFilteringAgentOptions) {
-        super(options);
-        applyRequestFilter(this, options);
     }
 }
 
 /**
  * A subclass of https.Agent with request filtering
  */
 export class RequestFilteringHttpsAgent extends https.Agent {
+    private requestFilterOptions: Required<RequestFilteringAgentOptions>;
+
     constructor(options?: https.AgentOptions & RequestFilteringAgentOptions) {
         super(options);
-        applyRequestFilter(this, options);
+        this.requestFilterOptions = {
+            allowPrivateIPAddress:
+                options && options.allowPrivateIPAddress !== undefined ? options.allowPrivateIPAddress : false,
+            allowMetaIPAddress:
+                options && options.allowMetaIPAddress !== undefined ? options.allowMetaIPAddress : false,
+            allowIPAddressList: options && options.allowIPAddressList ? options.allowIPAddressList : [],
+            denyIPAddressList: options && options.denyIPAddressList ? options.denyIPAddressList : []
+        };
+    }
+
+    // override http.Agent#createConnection
+    // https://nodejs.org/api/http.html#http_agent_createconnection_options_callback
+    // https://nodejs.org/api/net.html#net_net_createconnection_options_connectlistener
+    createConnection(options: TcpNetConnectOpts, connectionListener?: (error: Error | null, socket: Socket) => void) {
+        const { host } = options;
+        if (host !== undefined) {
+            // Direct ip address request without dns-lookup
+            // Example: http://127.0.0.1
+            // https://nodejs.org/api/net.html#net_socket_connect_options_connectlistener
+            const validationError = validateIPAddress({ address: host }, this.requestFilterOptions);
+            if (validationError) {
+                throw validationError;
+            }
+        }
+        // https://nodejs.org/api/net.html#net_socket_connect_options_connectlistener
+        // @ts-expect-error - @types/node does not defined createConnection
+        const socket: Socket = super.createConnection(options, connectionListener);
+        // Request with domain name
+        // Example: http://127.0.0.1.nip.io/
+        const onLookup = (err: Error, address: string, family: string | number, host: string): void => {
+            if (err) {
+                return;
+            }
+            const validationError = validateIPAddress({ address, family, host }, this.requestFilterOptions);
+            if (validationError) {
+                socket.removeListener("lookup", onLookup);
+                socket.end(() => {
+                    socket.destroy(validationError);
+                });
+            }
+        };
+        socket.addListener("lookup", onLookup);
+        return socket;
```

Comment:
Implement validation as Agent subclass.
Overriding instance(`applyRequestFilter`) cause ERR_INTERNAL_ASSERTION on Node.js

- https://github.com/nodejs/node/issues/48358

We need to avoid this error.